src/ioctls/sev.rs

```diff
@@ -0,0 +1,680 @@
+#![cfg(feature = "amd-sev")]
+
+use kvm_bindings::*;
+use std::fs::{File, OpenOptions};
+use std::mem::size_of_val;
+use std::os::unix::io::{AsRawFd, RawFd};
+
+use ioctls::{vm::VmFd, Result};
+use kvm_ioctls::*;
+use sev::launch::*;
+use vmm_sys_util::errno;
+use vmm_sys_util::ioctl::ioctl_with_mut_ref;
+
+#[repr(u32)]
+#[allow(dead_code)]
+#[derive(Copy, Clone, Debug)]
+pub enum SevCmd {
+    Init = 0,
+    EsInit,
+
+    LaunchStart,
+    LaunchUpdateData,
+    LaunchUpdateVmsa,
+    LaunchSecret,
+    LaunchMeasure,
+    LaunchFinish,
+
+    SendStart,
+    SendUpdateData,
+    SendUpdateVmsa,
+    SendFinish,
+
+    ReceiveStart,
+    ReceiveUpdateData,
+    ReceiveUpdateVmsa,
+    ReceiveFinish,
+
+    GuestStatus,
+    DebugDecrypt,
+    DebugEncrypt,
+    CertExport,
+}
+
+#[derive(Copy, Clone)]
+pub struct Handle(u32);
+/// Learn from ``SevState`` introduced in QEMU, adding few changes.
+/// An enumeration of SEV state information used during @query-sev.
+///
+/// @Uninit: The guest is uninitialized.
+///
+/// @Init: The guest is initialized.
+///
+/// @LaunchUpdate: The guest is currently being launched; plaintext data and
+///                register state is being imported.
+///
+/// @LaunchSecret: The guest is currently being launched; ciphertext data
+///                is being imported.
+///
+/// @Running: The guest is fully launched or migrated in.
+///
+/// @SendUpdate: The guest is currently being migrated out to another machine.
+///
+/// @ReceiveUpdate: The guest is currently being migrated from another machine.
+#[derive(PartialEq, Debug)]
+pub struct Uninit;
```

Comment:
This is meant to be an enumeration, so why not express it as an `Enum`? Also, modelling the state transitions with the `From` trait is a bit weird to me, and not very idiomatic IMO - `From` conceptually implies equivalence between the 2 operands, while a state machine implies not just non-equivalence, but order between the states. I believe an `Enum` here and explicit transition functions between the states would make the code clearer.

Reply:
> This is meant to be an enumeration, so why not express it as an `Enum`? Also, modelling the state transitions with the `From` trait is a bit weird to me, and not very idiomatic IMO - `From` conceptually implies equivalence between the 2 operands, while a state machine implies not just non-equivalence, but order between the states. I believe an `Enum` here and explicit transition functions between the states would make the code clearer.

Thanks for the suggestions. Seems reasonable to me. I'll make an update in the next version.